src/dbnode/persist/fs/merger.go

```diff
@@ -0,0 +1,287 @@
+	// persisted in the first stage.
+
+	// First stage: loop through series on disk.
+	for id, tagsIter, data, _, err := reader.Read(); err != io.EOF; id, tagsIter, data, _, err = reader.Read() {
+		idsToFinalize = append(idsToFinalize, id)
+		if err != nil {
+			return err
+		}
+
+		// Reset BlockReaders.
+		brs = brs[:0]
+		br = br[:0]
+		br = append(br, blockReaderFromData(data, segReader, startTime, blockSize))
+
+		// Check if this series is in memory (and thus requires merging).
+		mergeWithData, hasData, err := mergeWith.Read(id, blockStart, nsCtx)
+		if err != nil {
+			return err
+		}
+		if hasData {
+			br = append(br, mergeWithData...)
+		}
+		brs = append(brs, br)
+
+		sliceOfSlices.Reset(brs)
+		multiIter.ResetSliceOfSlices(sliceOfSlices, nsCtx.Schema)
+
+		// tagsIter is never nil.
+		tags, err := convert.TagsFromTagsIter(id, tagsIter, identPool)
+		tagsIter.Close()
+		tagsToFinalize = append(tagsToFinalize, tags)
+		if err != nil {
+			return err
+		}
+		if err := persistIter(prepared.Persist, multiIter, startTime,
+			id, tags, blockAllocSize, nsCtx.Schema, encoderPool); err != nil {
+			return err
+		}
+	}
+	// Second stage: loop through rest of the merge target that was not captured
+	// in the first stage.
+	err = mergeWith.ForEachRemaining(blockStart, func(seriesID ident.ID, tags ident.Tags) error {
+		mergeWithData, hasData, err := mergeWith.Read(seriesID, blockStart, nsCtx)
+		if err != nil {
+			return err
+		}
+
+		if hasData {
+			brs = brs[:0]
+			brs = append(brs, mergeWithData)
+			sliceOfSlices.Reset(brs)
+			multiIter.ResetSliceOfSlices(sliceOfSlices, nsCtx.Schema)
+			return persistIter(prepared.Persist, multiIter, startTime,
+				seriesID, tags, blockAllocSize, nsCtx.Schema, encoderPool)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	// Close the flush preparer, which writes the rest of the files in the
+	// fileset.
+	return prepared.Close()
+}
+
+func blockReaderFromData(
+	data checked.Bytes,
+	segReader xio.SegmentReader,
+	startTime time.Time,
+	blockSize time.Duration,
+) xio.BlockReader {
+	seg := ts.NewSegment(data, nil, ts.FinalizeHead)
+	segReader.Reset(seg)
+	return xio.BlockReader{
+		SegmentReader: segReader,
+		Start:         startTime,
+		BlockSize:     blockSize,
+	}
+}
+
+func persistIter(
+	persistFn persist.DataFn,
+	it encoding.Iterator,
+	blockStart time.Time,
+	id ident.ID,
+	tags ident.Tags,
+	blockAllocSize int,
+	schema namespace.SchemaDescr,
+	encoderPool encoding.EncoderPool,
+) error {
+	encoder := encoderPool.Get()
+	encoder.Reset(blockStart, blockAllocSize, schema)
+	for it.Next() {
+		if err := encoder.Encode(it.Current()); err != nil {
+			encoder.Close()
+			return err
+		}
+	}
+	if err := it.Err(); err != nil {
+		return err
+	}
+
+	segment := encoder.Discard()
+	checksum := digest.SegmentChecksum(segment)
+
+	err := persistFn(id, tags, segment, checksum)
+	if err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
nit: Perhaps combine this to just `return persistFn(id, tags, segment, checksum)`?